helios-client/src/main/java/com/spotify/helios/auth/AuthenticatingRequestDispatcher.java

```
@@ -0,0 +1,101 @@
+import static java.net.HttpURLConnection.HTTP_UNAUTHORIZED;
+
+public class AuthenticatingRequestDispatcher implements RequestDispatcher {
+
+  private final RequestDispatcher delegate;
+  private final AuthProvider authProvider;
+
+  public AuthenticatingRequestDispatcher(final RequestDispatcher delegate,
+                                         final AuthProvider authProvider) {
+    this.delegate = delegate;
+    this.authProvider = authProvider;
+  }
+
+  @Override
+  public ListenableFuture<Response> request(final HeliosRequest request) {
+    // Include an Authorization header if it's currently available
+    final HeliosRequest req;
+    final String authHeader = authProvider.currentAuthorization();
+    if (authHeader != null) {
+      req = request.toBuilder().header(HttpHeaders.AUTHORIZATION, authHeader).build();
+    } else {
+      req = request;
+    }
+
+    return Futures.transform(delegate.request(req), new AsyncFunction<Response, Response>() {
+      @Override
+      public ListenableFuture<Response> apply(final Response response) throws Exception {
+        if (response != null &&
+            response.status() == HTTP_UNAUTHORIZED &&
+            response.header(HttpHeaders.WWW_AUTHENTICATE) != null) {
+          return authenticateAndRetry(req, response);
+        } else {
+          return immediateFuture(response);
+        }
+      }
+    });
+  }
+
+  private ListenableFuture<Response> authenticateAndRetry(final HeliosRequest request,
+                                                          final Response response) {
+    final ListenableFuture<Response> f = Futures.transform(
+        authProvider.renewAuthorization(response.header(HttpHeaders.WWW_AUTHENTICATE)),
+        new AsyncFunction<String, Response>() {
+          @Override
+          public ListenableFuture<Response> apply(final String authHeader)
+              throws Exception {
+            return delegate.request(
```

I think that if the "new" token is the same as the previous one, we should return an error / not make a new request, assuming that the server will continue to think that value is not valid.